application.cpp

```diff
@@ -445,14 +462,18 @@ void application::exec() {
       (void)work;
       bool more = true;
       while( more || io_serv->run_one() ) {
+         if (my->_is_quiting)
+            break;
          while( io_serv->poll_one() ) {}
          // execute the highest priority item
          more = pri_queue.execute_highest();
       }
+      pri_queue.clear(); // make sure the queue is empty
 
       shutdown(); /// perform synchronous shutdown
    }
    io_serv.reset();
+   app_instance.reset(); // deleting *this... make sure it is the last thing we do.
```

Comment:
I think this might not interact well with exceptions? The idea seems to be that after `appbase::app().exec()` the next `appbase::app()` gets you a new `application` instance. But that isn't the case if an exception is thrown out of the io_context since this line will be skipped.

I feel like that might catch someone by surprise when they're writing a test.

Though, even in the original code it seems an exception would prevent `shutdown()` from being called? That seems pretty weird. So an exception that makes it out here causes plugins to be dtor'ed without `plugin_shutdown()`?

Reply:
> if an exception is thrown out of the io_context 

What do you mean by that? If an exception is thrown when executing `pri_queue.execute_highest();`?
Maybe we should have this in a try/catch block?

Reply:
Sloppy terminology on my part. Yeah an exception could be thrown by `run_one()`, `poll_one()` or `execute_highest()` I guess.

Reply:
I wonder if we should `catch`, log (cerr I guess), `shutdown`, and either rethrow or not. Probably not rethrow as `shutdown()` is a lot seems like re-throwing at that point is a bit too far from the exception.

Reply:
As @spoonincode pointed out, I think we have to rethrow the exception so main can return an error code.

Reply:
@heifner are you OK with the current state?

Reply:
I don't know what to think about the exception handling of shutdown() of plugins. If we keep it as you have it now I think it should throw the first exception caught, not the last exception. I wonder if it is better to not catch exceptions in plugin_shutdown() and make that a responsibility of the plugins themselves.

Reply:
If the plugin themselves catch exceptions (as they should), then what we do here doesn't matter very much.

If one plugin does *not* catch exception abd throws, I think it is very important we give the other plugins a chance to shutdown correctly. It may make the difference between leaving a corrupted database on disk ve a clean one.

Reply:
I believe this conversation is resolved!